<a href="https://colab.research.google.com/github/ahsan-shakeel/Flyrank_ML_internship/blob/main/Flyrank_ML_internship/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Chosen Lane: Refresh / Content Opportunity Scoring

Why: We have 500k+ content items across multiple clients, making manual inspection impossible. This lane helps us prioritize which decaying or underperforming pages require content updates first, maximizing content team ROI.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Unit of Analysis (Grain): A single content item (content_hash_id).

Decision: Deciding which pages should be sent to the content review queue.

Action: Content team expands thin content, updates stale facts, or optimizes title/meta descriptions.

Cost of Wrong Call:

False Positive (Flagging a healthy page): Waste of 2–4 hours of manual editor capacity.

False Negative (Missing a decaying page): Permanent loss of search traffic, rankings, and revenue.

Why ML/Data: Simple age rules ("update pages > 180 days old") fail because old pages often hold rank while new pages decay. ML handles multi-variable patterns (traffic trend + position decay + impression volume) simultaneously.

In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

The previous cell encountered a `FileNotFoundError`. This is because the CSV file `content_refresh_anonymized.csv` was not present in the expected `data/raw` directory. I will add the necessary commands to download this file from the project's GitHub repository and create the directory if it doesn't exist.

In [13]:
# Create the 'data/raw' directory if it doesn't exist
!mkdir -p data/raw

# Download the CSV file from the GitHub repository
!wget -O data/raw/content_refresh_anonymized.csv https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv

--2026-08-05 12:08:00--  https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6727670 (6.4M) [text/plain]
Saving to: ‘data/raw/content_refresh_anonymized.csv’

data/raw/content_re 100%[===================>]   6.42M  25.2MB/s    in 0.3s    

2026-08-05 12:08:00 (25.2 MB/s) - ‘data/raw/content_refresh_anonymized.csv’ saved [6727670/6727670]



In [14]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

# Load starter dataset
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Extract 3 real numbers to back your lane
total_pages = len(df)
declining_pages = (df['trend_direction'] == 'down').sum()
decline_pct = (declining_pages / total_pages) * 100
median_impressions = df['impressions_90d'].median()

print(f"1. Total Pages in Starter Slice: {total_pages:,}")
print(f"2. Declining Pages (trend_direction == 'down'): {declining_pages:,} ({decline_pct:.1f}%)")
print(f"3. Median 90-Day Impressions: {median_impressions:.0f}")

1. Total Pages in Starter Slice: 30,000
2. Declining Pages (trend_direction == 'down'): 16,262 (54.2%)
3. Median 90-Day Impressions: 731


Starter dataset contains 30,000 unique pages, giving a large sample to rank.

54.2% of pages (approx 16,262 pages) are currently experiencing a downward traffic trend (trend_direction == 'down').

The median 90-day impressions count is 731, showing sufficient demand volume to filter out low-volume noise.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [15]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


What I CAN Claim: "The model generates an evidence-backed priority score to rank pages for review based on historical traffic, CTR, and position signals."

What I CANNOT Claim: "I cannot claim that refreshing a page guarantees a Google rank recovery, nor that the model proves Google algorithm factors or causal impact without an A/B experiment."

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.